src/credentials/CHIPCert.cpp

```diff
@@ -0,0 +1,831 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements objects for modeling and working with
+ *      CHIP certificates.
+ *
+ */
+
+#ifndef __STDC_LIMIT_MACROS
+#define __STDC_LIMIT_MACROS
+#endif
+
+#include <stddef.h>
+
+#include <asn1/ASN1.h>
+#include <asn1/ASN1Macros.h>
+#include <core/CHIPCore.h>
+#include <core/CHIPSafeCasts.h>
+#include <core/CHIPTLV.h>
+#include <credentials/CHIPCert.h>
+#include <protocols/Protocols.h>
+#include <support/CHIPMem.h>
+#include <support/CodeUtils.h>
+#include <support/TimeUtils.h>
+
+namespace chip {
+namespace Credentials {
+
+using namespace chip::ASN1;
+using namespace chip::TLV;
+using namespace chip::Protocols;
+using namespace chip::Crypto;
+
+extern CHIP_ERROR DecodeConvertTBSCert(TLVReader & reader, ASN1Writer & writer, ChipCertificateData & certData);
+
+ChipCertificateSet::ChipCertificateSet()
+{
+    Certs                = nullptr;
+    CertCount            = 0;
+    MaxCerts             = 0;
+    mDecodeBuf           = nullptr;
+    mDecodeBufSize       = 0;
+    mMemoryAllocInternal = false;
+}
+
+CHIP_ERROR ChipCertificateSet::Init(uint8_t maxCertsArraySize, uint16_t decodeBufSize)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    VerifyOrExit(maxCertsArraySize > 0, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(decodeBufSize > 0, err = CHIP_ERROR_INVALID_ARGUMENT);
+
+    Certs = reinterpret_cast<ChipCertificateData *>(chip::Platform::MemoryAlloc(sizeof(ChipCertificateData) * maxCertsArraySize));
+    VerifyOrExit(Certs != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    CertCount            = 0;
+    MaxCerts             = maxCertsArraySize;
+    mDecodeBuf           = nullptr;
+    mDecodeBufSize       = decodeBufSize;
+    mMemoryAllocInternal = true;
+
+exit:
+    return err;
+}
+
+CHIP_ERROR ChipCertificateSet::Init(ChipCertificateData * certsArray, uint8_t certsArraySize, uint8_t * decodeBuf,
+                                    uint16_t decodeBufSize)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    VerifyOrExit(certsArray != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(certsArraySize > 0, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(decodeBuf != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(decodeBufSize > 0, err = CHIP_ERROR_INVALID_ARGUMENT);
+
+    CertCount            = 0;
+    Certs                = certsArray;
+    MaxCerts             = certsArraySize;
+    mDecodeBuf           = decodeBuf;
+    mDecodeBufSize       = decodeBufSize;
+    mMemoryAllocInternal = false;
+
+exit:
+    return err;
+}
+
+void ChipCertificateSet::Release()
+{
+    if (mMemoryAllocInternal)
+    {
+        if (Certs != nullptr)
+        {
+            chip::Platform::MemoryFree(Certs);
+            Certs = nullptr;
+        }
+        if (mDecodeBuf != nullptr)
+        {
+            chip::Platform::MemoryFree(mDecodeBuf);
```

Comment:
is mDecodeBuf ever set in the mMemoryAllocInternal case?

Reply:
now it is.